### PR TITLE
Added string to pg_float8 typecasting

### PR DESCRIPTION
**Pull request by mohayu22**
**Why:**

While it is a common practice to convert a _string_ to an _integer_ or _float_ if it resembles one, this functionality was missing for `pg_float8`
```
SELECT * from cypher('test', $$
    RETURN "1.23"::pg_float8
$$) as (result agtype);

ERROR:  cannot cast agtype string to type float
```
**Solution:**

- Modified `agtype_to_float8` function
- Modified the if condition to check for _AGTV_STRING_
- Added the code for case _AGTV_STRING_

### CODE REVIEW

**Comment by mohayu22** [comment]
@jrgemignani checks are pending for this PR

**Comment by jrgemignani** [comment]
@mohayu22 Please add in more regression tests to test all potential cases and edge cases.

**Comment by mohayu22** [comment]
@jrgemignani While testing for potential cases to be added in this PR, I discovered that the code works correctly for all the new cases except for the function agtype_in. The reason it was failing is that this function only checks for the types `::integer`, `::float`, and `::numeric`. It does not handle `::int`, `::pg_float8`, or `::pg_bigint`.

```
SELECT agtype_in('"124"::integer');
 agtype_in 
-----------
 124
(1 row)

SELECT agtype_in('"124"::int');
ERROR:  invalid annotation value for scalar

SELECT agtype_in('"124"::pg_bigint');
ERROR:  invalid annotation value for scalar

SELECT agtype_in('"12.4"::float');
 agtype_in 
-----------
 12.4
(1 row)

SELECT agtype_in('"12.4"::pg_float8');
ERROR:  invalid annotation value for scalar
```

**Comment by jrgemignani** [comment]
@mohayu22 That is fine. We just need to make sure that we cover all the cases in regression tests, good and bad. Above the test cases we usually state whether it should pass or fail, sometimes why, sometimes how many rows selected, etc. More information is usually better as it helps if those regression tests have issues in the future.

File: regress/expected/expr.out [hunk 1]
```diff
@@ -1667,6 +1667,135 @@ SELECT agtype_in('[NaN, {"e": 2.718281::numeric, "one": Infinity, "pie": 3.14159
...
+ERROR:  invalid input syntax for type double precision: ""
+SELECT * FROM cypher('expr', $$
+RETURN '2:71'::pg_float8
+$$) AS r(result agtype);
+ERROR:  invalid input syntax for type double precision: "2:71"
+SELECT * FROM cypher('expr', $$
+RETURN 'infi'::pg_float8
+$$) AS r(result agtype);
+ERROR:  invalid input syntax for type double precision: "infi"
+SELECT * FROM cypher('expr', $$
```
Comment (jrgemignani) [inline regress/expected/expr.out:1791]: Should this fail or does it need to be in a different group?

File: regress/expected/expr.out [hunk 2]
```diff
@@ -1667,6 +1667,135 @@ SELECT agtype_in('[NaN, {"e": 2.718281::numeric, "one": Infinity, "pie": 3.14159
...
+RETURN ([0, {one: 1::pg_float8, pie: 3.1415927, e: 2.718281::numeric}, 2, null][1].one)::float
+$$) AS r(result agtype);
+ result 
+--------
+ 1.0
+(1 row)
+
+SELECT * FROM cypher('expr', $$
+RETURN ([0, {one: 1, pie: 3.1415927, e: 2.718281::numeric}, 2, null][3])::pg_float8
```
Comment (jrgemignani) [inline regress/expected/expr.out:1722]: what about casting a numeric?

**Review by jrgemignani — CHANGES_REQUESTED** [review]

File: regress/expected/expr.out (hunk 1 above)
Comment (mohayu22) [reply regress/expected/expr.out:1799]: Yes, it should fail and it is in the right block

File: regress/expected/expr.out (hunk 2 above)
Comment (mohayu22) [reply regress/expected/expr.out:1722]: Added

File: regress/expected/expr.out (hunk 1 above)
Comment (mohayu22) [reply regress/expected/expr.out]: My bad, the test below this "infi" one was not failing and was in the wrong category, moved it up.

**Review by jrgemignani — APPROVED** [review]
Normally I would have asked you to make sure that all if/else structures have {} even for 1 line of code. However, there is another patch to the same function that addresses these.

**Comment by jrgemignani** [comment]
@mohayu22 Could you update your PR against the master. I just approved and merged another PR that made modifications to the same code and I want to be sure all is good.

**Comment by mohayu22** [comment]
@jrgemignani Done

File: src/backend/utils/adt/agtype.c [hunk 3]
```diff
@@ -2788,6 +2789,9 @@ Datum agtype_to_float8(PG_FUNCTION_ARGS)
     else if (agtv.type == AGTV_NUMERIC)
         result = DatumGetFloat8(DirectFunctionCall1(numeric_float8,
                      NumericGetDatum(agtv.val.numeric)));
+    else if (agtv.type == AGTV_STRING)
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2792]: All if/else blocks need to have {} for all code, even if it is only one line.

File: src/backend/utils/adt/agtype.c [hunk 4]
```diff
@@ -2788,6 +2789,9 @@ Datum agtype_to_float8(PG_FUNCTION_ARGS)
...
+        result = DatumGetFloat8(DirectFunctionCall1(float8in,
+                                                    CStringGetDatum(agtv.val.string.val)));
     else
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2795]: When we work on functions that don't conform to our standards, we proactively try to clean them up. Just the functions that we are working on. If you could adjust the function to have {} where necessary.

File: src/backend/utils/adt/agtype.c [hunk 5]
```diff
@@ -2788,6 +2789,9 @@ Datum agtype_to_float8(PG_FUNCTION_ARGS)
     else if (agtv.type == AGTV_NUMERIC)
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2789]: When we work on functions that don't conform to our standards, we proactively try to clean them up. Just the functions that we are working on. If you could adjust the function to have {} where necessary.

File: src/backend/utils/adt/agtype.c [hunk 6]
```diff
@@ -2759,7 +2759,8 @@ Datum agtype_to_float8(PG_FUNCTION_ARGS)
     if (!agtype_extract_scalar(&agtype_in->root, &agtv) ||
         (agtv.type != AGTV_FLOAT &&
          agtv.type != AGTV_INTEGER &&
-         agtv.type != AGTV_NUMERIC))
+         agtv.type != AGTV_NUMERIC &&
+         agtv.type != AGTV_STRING))
```
Comment (jrgemignani) [inline src/backend/utils/adt/agtype.c:2763]: When we work on functions that don't conform to our standards, we proactively try to clean them up. Just the functions that we are working on. If you could adjust the function to have {} where necessary.

**Review by jrgemignani — CHANGES_REQUESTED** [review]
Minor code adjustments.

**Review by jrgemignani — APPROVED** [review]

**Comment by jrgemignani** [comment]
@mohayu22 Good job!

**Comment by jrgemignani** [comment]
@mohayu22 For future reference, your commit description block should be more concise -

```
commit a43002ae39132781c535031d426d04bc5464851e
Author: mohayu22 <113895061+mohayu22@users.noreply.github.com>
Date:   Wed May 10 23:25:54 2023 +0500

    Added string to pg_float8 typecasting (#895)

    * Added case for AGTV_STRING

    -Made changes to agtype_to_float8
    -Modified if condition to check for AGTV_STRING
    -Added code for typecasting string to pg_float8

    * Added and modified regression tests

    -Modified a test to return value instead of failing
    -Added new test to check string to pg_flaot8 typecasting

    * Added braces around code blocks

    Added braces {} around code blocks to keep up with standards.
```

Could be more like - 

```
commit a43002ae39132781c535031d426d04bc5464851e
Author: mohayu22 <113895061+mohayu22@users.noreply.github.com>
Date:   Wed May 10 23:25:54 2023 +0500

    Added string to pg_float8 typecasting (#895)

    * Added case for AGTV_STRING

    -Made changes to agtype_to_float8
    -Modified if condition to check for AGTV_STRING
    -Added code for typecasting string to pg_float8

    * Added and modified regression tests
```
